File: packages/gms_places_api/android/src/main/kotlin/dev/laihz/gms_places_api/GmsPlacesApiImpl.kt
package dev.laihz.gms_places_api

import android.content.Context
import com.google.android.libraries.places.api.Places
import com.google.android.libraries.places.api.model.AutocompleteSessionToken
import com.google.android.libraries.places.api.model.PlaceTypes
import com.google.android.libraries.places.api.net.FindAutocompletePredictionsRequest

class GmsPlacesApiImpl(private val context: Context) : GmsPlacesApi {
    companion object {
        var token: AutocompleteSessionToken? = null
    }

    override fun ensureInitialized() {
        token = AutocompleteSessionToken.newInstance()
    }

    override fun autocomplete(
        fromQuery: String,
        filter: PlacesFilterType,
        callback: (Result<List<Prediction>>) -> Unit
    ) {
        val filter = when (filter) {
            PlacesFilterType.GEOCODE -> PlaceTypes.GEOCODE
            PlacesFilterType.ADDRESS -> PlaceTypes.ADDRESS
            PlacesFilterType.ESTABLISHMENT -> PlaceTypes.ESTABLISHMENT
            PlacesFilterType.REGION -> PlaceTypes.REGIONS
            PlacesFilterType.CITY -> PlaceTypes.ADDRESS
        }
        val req = FindAutocompletePredictionsRequest.builder()
            .setTypesFilter(listOf(filter))
            .setSessionToken(token)
            .setQuery(fromQuery)
            .build()

        Places.createClient(context).findAutocompletePredictions(req)
            .addOnSuccessListener {response ->
                val items = mutableListOf<Prediction>()
                for (item in response.autocompletePredictions) {
                    items.add(
                        Prediction(
                        attributed = PredictionAttributed(
                            fullText = item.getFullText(null).toString(),
                            primaryText = item.getPrimaryText(null).toString(),
                            secondaryText = item.getSecondaryText(null).toString(),
                        ),
                        placeID = item.placeId,
                        rawTypes = item.placeTypes.map { it -> it.name }
                    )
                    )
                }
                callback(Result.success(items))
            }
            .addOnFailureListener {exception ->
                callback(Result.failure(exception))
            }
    }


}

File: packages/gms_places_api/ios/Classes/GmsPlacesApiImpl.swift
import Flutter
import GooglePlaces

extension FlutterError: Error {}

class GmsPlacesApiImpl: GmsPlacesApi {
    static var token: GMSAutocompleteSessionToken?

    func ensureInitialized() throws {
        GmsPlacesApiImpl.token = GMSAutocompleteSessionToken()
    }

    func autocomplete(fromQuery: String, filter: PlacesFilterType, completion: @escaping (Result<[Prediction], Error>) -> Void) {
        let autocompleteFilter = GMSAutocompleteFilter()
        switch filter {
        case .address:
            autocompleteFilter.type = .address
        case .city:
            autocompleteFilter.type = .city
        case .establishment:
            autocompleteFilter.type = .establishment
        case .geocode:
            autocompleteFilter.type = .geocode
        case .region:
            autocompleteFilter.type = .region
        }
        if GmsPlacesApiImpl.token == nil {
            let error = FlutterError(code: "emptyToken", message: "emptyToken", details: "emptyToken")
            completion(.failure(error))
        }
        GMSPlacesClient().findAutocompletePredictions(
            fromQuery: fromQuery,
            filter: autocompleteFilter,
            sessionToken: GmsPlacesApiImpl.token
        ) { items, error in
            if error != nil {
                completion(.failure(FlutterError(code: "searchError", message: error?.localizedDescription, details: "")))
            } else {
                if items == nil {
                    completion(.success([]))
                    return
                }
                var result = [Prediction]()
                for item in items! {
                    result.append(Prediction(
                        attributed: PredictionAttributed(
                            fullText: item.attributedFullText.string,
                            primaryText: item.attributedPrimaryText.string,
                            secondaryText: item.attributedSecondaryText?.string
                        ),
                        placeID: item.placeID,
                        rawTypes: item.types,
                        distanceMeters: item.distanceMeters?.doubleValue
                    ))
                }
                completion(.success(result))
            }
        }
    }
}

class TokenEmptyError: Error {}
